### ISSUE TITLE
Questions about programatic use of the Service Broker

### ISSUE THREAD

**Issue by marpaia** [kubernetes-incubator/service-catalog#1943]
As I was discussing in Slack with @carolynvs, I maintain a `Tenant` CRD and controller. The Tenant resource allows you to define application resources to deploy (templated kubernetes yaml smarty, helm charts, ksonnet, etc) as well as a set of OSB Service Instances that you need to bind to. Consider the example resource [here](https://gist.github.com/marpaia/d2d88a35574d42c9b1af27e1eeb21d80).

The CRD basically defines an abstraction on top of Service Catalog that allows you to more easily deploy many instances of an application on a cluster. For example, consider the `spec.serviceCatalog.instances` section of the CRD:

```yaml
spec:
  serviceCatalog:
    instances:
      - name: mysql
        clusterServiceClassExternalName: azure-mysql
        clusterServicePlanExternalName: basic50
        parameters:
          location: eastus
```

Each item in the `instances` section is analagous to a `ServiceInstace` type from the `github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog` package. Given each instance, my controller will take care of interacting with the Service Catalog API server to create/install the binding in the appropriate namespace, etc.

From some brief spelunking, if I want to manage service catalog resources from my controller, I would need two components as prerequisites:

- an instance of the API server
  - `runEtcdServer` in `cmd/apiserver/app/server/run_server.go` is the most succinct example of that I've found
- a Controller which operates on the service catalog resources
  - entrypoint: `github.com/kubernetes-incubator/service-catalog/pkg/controller.NewController`
  - requires the API server to exist already (because a `kubeClient` to that API server is a dependency)
  - controller uses normal shared informer workqueue controller tools to process resources

Given this, if I start an API server on one port in my controller (maybe not even binding to `0.0.0.0`), then I can create a Service Catalog controller alongside the controller that I'm already running for my CRDs.

How does that sound? Sane? I know y'all are considering replacing the `apiserver` with CRDs (and maybe the admission controller webhook things?). Reading through the `pkg/apiserver` code, that doesn't look like a small task though...

Also, would there be anything to consider if a user of my controller installed service catalog via the `APIServer` resource (as in the helm chat)?

**Comment by n3wscott** [comment]
why not just install service catalog directly and your CRDs help manage the service catalog resources?

**Comment by marpaia** [comment]
That's more or less what I'm proposing, but right now the installation UX for my CRDs+controller is:

```
curl -L https://lessor.io/latest/lessor.yaml | kubectl apply -f -
```

Given that the current installation instructions for Service Catalog require installing Helm first, that seems like a blocker for my use-case, so I'm trying to get my head around what is required to be able to operate on the service catalog resources in my own kind of isolated/namespaced way.

**Comment by marpaia** [comment]
I agree with you though @n3wscott, if there is a minimal, embeddable, easily distributable way for me to just get a client to the service catalog API server (or the main API server with CRDs), then that would be ideal.

**Comment by marpaia** [comment]
I could take a run at making a hosted manifest out of the [Helm templates](https://github.com/kubernetes-incubator/service-catalog/tree/master/charts/catalog/templates) since it doesn't seem like there is much I'd want to change from installation to installation?

**Comment by MHBauer** [comment]
I really like your big comment on the `serviceCatalog:` field in your example yaml. Especially [this bit](https://gist.github.com/marpaia/d2d88a35574d42c9b1af27e1eeb21d80#file-tenant-yaml-L23-L25)

Go client is in https://github.com/kubernetes-incubator/service-catalog/tree/master/pkg/client

I feel from your description that the binding is something we also want to be doing, we just haven't gotten there yet. We were going to take advantage of podpresets. Without automated binding injection, we really only serve as a semi-convenient central place for people to look at services.

**Comment by marpaia** [comment]
@MHBauer it's my understanding that if I create a binding for a service and plan in my catalog (in namespace `foo`), Service Catalog will create a secret in namespace `foo` with the credentials it got in the response from `PUT /v2/service_instances/:instance_id/service_bindings/:binding_id`. Is that what you're talking about?

**Comment by MHBauer** [comment]
An original goal is to do more than that. Take the credentials and put them in a secret, yes, but also create a podpreset that automatically injects the contents of the secret into pods that fullfill a selector.

**Comment by marpaia** [comment]
Oh, very cool. I think that would be above and beyond.

**Comment by carolynvs** [comment]
I misunderstood what you were hoping to do when we chatted on slack. I thought the question was "should I talk directly to OSB or to service catalog", which is why I thought exposing any of our client-side functionality better would be useful to you.

But it sounds like you actually want to completely host, wrap/proxy service catalog yourself?

**Comment by marpaia** [comment]
@carolynvs yeah, service catalog is a dependency of my project. so the user *may* have already installed it, but if they don't, they can't use my thing unless they have service catalog. So I need to include it and the strategy that seems the most sane is to use the Helm chat to produce a manifest with my suggested installation locations and configurations and bundle it.

Ideally service catalog would just exist in upstream Kubernetes. If I couldn't have that, just telling users to run `svcat install` would be OK. So it seems like better solutions are on the current path of development anyway.

**Comment by carolynvs** [comment]
Ah okay, sorry about that. I'm glad that `svcat install catalog` (#1932) would work for you as that's definitely on our radar.
